pkg/autoscaler/aggregation/bucketing.go

```diff
@@ -87,6 +108,49 @@ func roundToNDigits(n int, f float64) float64 {
 	return math.Floor(f*p) / p
 }
 
+const (
+	// minExponent is the minimal decay multiplier for the weighted average computations.
+	minExponent = 0.2
+	// The sum of weights for weighted average should be at least this much.
+	weightPrecision = 0.9999
+	precision       = 6
+)
+
+// WeightedAverage returns the exponential weighted average
+// TODO(vagababov): optimize for O(1) computation, if possible.
+func (t *TimedFloat64Buckets) WeightedAverage(now time.Time) float64 {
```

Comment:
Sort of wondering if it'd be better to pull this in to its own file/struct, at least for now (i.e. `WeightedTimedFloat64Buckets`). TimedFloat64Buckets is already pretty complicated without two different averaging methods to worry about. That way we can keep the same contract for both and just use an interface for them when we use them so scaler doesn't need to know about weighting. Does mean duplicating Record, but I wouldn't be surprised if we needed to change that to get closer to O(1) for this anyway?

Reply:
This might be a better longer term strategy if we decide to keep this feature. I am less concerned about target metric change, this should be rare and it's brewing in my head that we should make metric immutable.